src/main/java/org/jvnet/hudson/test/RestartableJenkinsRule.java

```diff
@@ -76,6 +83,63 @@ public void evaluate() throws Throwable {
         });
     }
 
+    /** Approach adapted from https://stackoverflow.com/questions/6214703/copy-entire-directory-contents-to-another-directory */
+    static class CopyFileVisitor extends SimpleFileVisitor<Path> {
+        private final Path targetPath;
+        private Path sourcePath = null;
+        public CopyFileVisitor(Path targetPath) {
+            this.targetPath = targetPath;
+        }
+
+        @Override
+        public FileVisitResult preVisitDirectory(final Path dir,
+                                                 final BasicFileAttributes attrs) throws IOException {
+            if (sourcePath == null) {
+                sourcePath = dir;
+            } else {
+                Files.createDirectories(targetPath.resolve(sourcePath
+                        .relativize(dir)));
+            }
+            return FileVisitResult.CONTINUE;
+        }
+
+        @Override
+        public FileVisitResult visitFile(final Path file,
+                                         final BasicFileAttributes attrs) throws IOException {
+            if (!Files.isSymbolicLink(file)) {
+                // Needed because Jenkins includes invalid lastSuccessful symlinks and otherwise we get a NoSuchFileException
+                Files.copy(file,
+                        targetPath.resolve(sourcePath.relativize(file)));
+            } else if (Files.isSymbolicLink(file) && Files.exists(Files.readSymbolicLink(file))) {
+                Files.copy(file,
+                        targetPath.resolve(sourcePath.relativize(file)));
+            }
+            return FileVisitResult.CONTINUE;
+        }
+    }
+
+    /**
+     * Simulate an abrupt failure of Jenkins to see if it appropriately handles inconsistent states when
+     *  shutdown cleanup is not performed or data is not written fully to disk.
+     *
+     * Works by copying the JENKINS_HOME to a new directory and then setting the {@link RestartableJenkinsRule} to use
+     * that for the next restart. Thus we only have the data actually persisted to disk at that time to work with.
+     *
+     * Should be run as the last part of a {@link org.jvnet.hudson.test.RestartableJenkinsRule.Step}.
+     *
+     * @throws IOException
+     */
+     void simulateAbruptShutdown() throws IOException {
+        File homeDir = this.home;
+        TemporaryFolder temp = new TemporaryFolder();
+        temp.create();
+        File newHome = temp.newFolder();
+
+        // Copy efficiently
+        Files.walkFileTree(homeDir.toPath(), Collections.EMPTY_SET, 99, new CopyFileVisitor(newHome.toPath()));
```

Comment:
Why not just use `FilePath.copyRecursiveTo`, and if that is not as fast as you think it could be, optimize it? Surely we do not need blazing performance for a test utility anyway.

Reply:
+1, also would be indeed an opportunity to actually have a bigger perf impact, if need be. Since `copyRecursiveTo` is used in Pipeline too FWICT by the `libraryAdder`.